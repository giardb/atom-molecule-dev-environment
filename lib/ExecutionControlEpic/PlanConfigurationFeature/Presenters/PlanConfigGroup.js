"use babel";
// @flow

import React from "react";
import Image from "react-image-fallback";
import type { ConfigSchemaPart } from "../Types/types.js.flow";
import PlanConfigPart from "./PlanConfigPart";
import styled from "styled-components";
// $FlowFixMe
import { StateProxy } from "react-forms-state";
import { compose, withState } from "recompose";

const Box = styled.div`
  display: flex;
  flex-direction: column;
  align-items: stretch;
  flex-shrink: 0;
  margin: 8px 0px;
`;

const LabelBox = styled.div`
  display: flex;
  justify-content: space-between;
  flex-shrink: 0;
  margin: 4px 0px;
`;

const Label = styled.span.attrs({
  className: "text-color-highlight",
})`
  display: flex;
  font-size: 13px;
`;

const ContentBox = styled.div`
  display: flex;
  flex-direction: column;
  flex-shrink: 0;
  align-items: stretch;
`;

export function PlanConfigGroup({
  label,
  expanded = true,
  setExpanded = () => {},
  schemas = {},
}: {
  label: string,
  expanded: boolean,
  setExpanded: (v: boolean) => void,
  schemas: { [key: string]: ConfigSchemaPart },
  value: any,
  onChange: (v: any) => void,
}) {
  return (
    <Box>
      <LabelBox onClick={() => setExpanded(!expanded)}>
        <Label>
          config{label != null ? ` > ${label}` : ""}
        </Label>
        <Image
          src="atom://molecule-dev-environment/.storybook/public/arrow-right.svg"
          fallbackImage="arrow-right.svg"
          style={{
            display: "flex",
            height: "16px",
            width: "16px",
            transform: expanded ? "rotate(90deg)" : "",
            transition: "0.2s transform",
          }}
        />
      </LabelBox>
      {expanded &&
        <ContentBox>
          {Object.keys(schemas).map(key =>
            <PlanConfigPart {...schemas[key]} name={key} key={key} />,
          )}
        </ContentBox>}
    </Box>
  );
}

const enhance = compose(
  StateProxy(),
  withState("expanded", "setExpanded", true),
);

export default enhance(PlanConfigGroup);
