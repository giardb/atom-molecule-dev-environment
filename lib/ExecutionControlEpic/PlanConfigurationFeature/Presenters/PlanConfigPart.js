"use babel";
// @flow

import React from "react";
import PlanConfigSelectInputField from "./PlanConfigSelectInputField";
import PlanConfigTextInputField from "./PlanConfigTextInputField";
import PlanConfigGroup from "./PlanConfigGroup";
import type {
  ArrayConfigSchema,
  ConditionalConfigSchema,
  ConfigSchemaPartBase,
  EnumConfigSchema,
  NumberConfigSchema,
  ObjectConfigSchema,
  StringConfigSchema,
  FileConfigSchema,
} from "../Types/types.js.flow";
import PlanConfigConditionalGroup from "./PlanConfigConditionalGroup";
import PlanConfigNumberInputField from "./PlanConfigNumberInputField";
import PlanConfigCheckboxInputField from "./PlanConfigCheckboxInputField";
import PlanConfigFileInputField from "../Containers/PlanConfigFileInputField";
import PlanConfigList from "./PlanConfigList";

export default function PlanConfigPart({
  type,
  value,
  onChange,
  label,
  items,
  expression,
  cases,
  placeholder,
  schemas,
  tester,
  ...props
}: ConfigSchemaPartBase & {
  onChange: (value: any) => void,
  // eslint-disable-next-line prettier/prettier
} & ( // eslint-disable-next-line prettier/prettier
    | (EnumConfigSchema & { value: any })
    | (FileConfigSchema & { value: any })
    | (ObjectConfigSchema & { value: { [key: string]: any } })
    | (StringConfigSchema & { value: string })
    | (NumberConfigSchema & { value: number })
    | (ConditionalConfigSchema & {
      value: { expressionValue: mixed, caseValue: mixed },
    })
    | (ArrayConfigSchema & { value: Array<any> }))) {
  switch (type) {
    case "enum":
      return (
        <PlanConfigSelectInputField
          options={props.enum}
          default={props.default}
          value={value}
          onChange={onChange}
        />
      );
    case "file":
      return (
        <PlanConfigFileInputField
          tester={tester}
          default={props.default}
          value={value}
          label={label}
          onChange={onChange}
        />
      );
    case "object":
      return (
        <PlanConfigGroup
          label={label}
          schemas={schemas}
          value={value}
          onChange={onChange}
        />
      );
    case "string":
      return (
        <PlanConfigTextInputField
          default={props.default}
          value={value}
          onChange={onChange}
          label={label}
          placeholder={placeholder}
        />
      );
    case "number":
      return (
        <PlanConfigNumberInputField
          default={props.default}
          value={value}
          onChange={onChange}
          label={label}
        />
      );
    case "conditional":
      return (
        <PlanConfigConditionalGroup
          default={props.default}
          expression={expression}
          cases={cases}
          value={value}
          onChange={onChange}
        />
      );
    case "array":
      return (
        <PlanConfigList
          values={value}
          default={props.default}
          onChange={onChange}
          items={items}
          label={label}
        />
      );
    case "boolean":
      return (
        <PlanConfigCheckboxInputField
          value={value}
          default={props.default}
          onChange={onChange}
          label={label}
        />
      );
    default:
      return null;
  }
}
