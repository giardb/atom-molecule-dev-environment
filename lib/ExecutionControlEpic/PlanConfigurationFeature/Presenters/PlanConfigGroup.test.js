import React from "react";
import { mount } from "enzyme";
import { PlanConfigGroup } from "./PlanConfigGroup";
import PlanConfigPart from "./PlanConfigPart";

describe("PlanConfigGroup", () => {
  it("should display a PlanConfigPart for every key of schemas", () => {
    let subject = mount(
      <PlanConfigGroup
        schemas={{
          image: { type: "string" },
          scale: { type: "number" },
        }}
        value={{ image: "ubuntu", scale: 3 }}
      />,
    );

    expect(subject.find(PlanConfigPart).length).toBe(2);
    expect(subject.find(PlanConfigPart).at(0).prop("type")).toBe("string");
    expect(subject.find(PlanConfigPart).at(0).prop("value")).toBe("ubuntu");
    expect(subject.find(PlanConfigPart).at(1).prop("type")).toBe("number");
  });

  it("should call onChange when a part is changed", () => {
    let spy = jest.fn();
    let subject = mount(
      <PlanConfigGroup
        schemas={{
          image: { type: "string" },
          scale: { type: "number" },
        }}
        onChange={spy}
        value={{ image: "ubuntu", scale: 3 }}
      />,
    );

    subject.find("PlanConfigPart").at(1).prop("onChange")(4);

    expect(spy).toBeCalledWith({ image: "ubuntu", scale: 4 });
  });
});
