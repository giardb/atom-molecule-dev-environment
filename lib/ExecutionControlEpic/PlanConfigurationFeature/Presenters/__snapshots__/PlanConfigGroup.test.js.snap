// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanConfigGroup should display a PlanConfigPart for every key of schemas 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <styled.div>
    <styled.div
        onClick={[Function]}
    >
        <styled.span>
            config
            
        </styled.span>
        <ReactImageFallback
            fallbackImage="arrow-right.svg"
            initialImage={null}
            src="atom://molecule-dev-environment/.storybook/public/arrow-right.svg"
            style={
                Object {
                    "display": "flex",
                    "height": "16px",
                    "transform": "rotate(90deg)",
                    "transition": "0.2s transform",
                    "width": "16px",
                  }
            }
        />
    </styled.div>
    <styled.div>
        <PlanConfigPart
            name="image"
            type="string"
        />
        <PlanConfigPart
            name="scale"
            type="number"
        />
    </styled.div>
</styled.div>,
  "nodes": Array [
    <styled.div>
      <styled.div
            onClick={[Function]}
      >
            <styled.span>
                  config
                  
            </styled.span>
            <ReactImageFallback
                  fallbackImage="arrow-right.svg"
                  initialImage={null}
                  src="atom://molecule-dev-environment/.storybook/public/arrow-right.svg"
                  style={
                        Object {
                              "display": "flex",
                              "height": "16px",
                              "transform": "rotate(90deg)",
                              "transition": "0.2s transform",
                              "width": "16px",
                            }
                  }
            />
      </styled.div>
      <styled.div>
            <PlanConfigPart
                  name="image"
                  type="string"
            />
            <PlanConfigPart
                  name="scale"
                  type="number"
            />
      </styled.div>
</styled.div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <PlanConfigGroup
        schemas={
                Object {
                        "image": Object {
                          "type": "string",
                        },
                        "scale": Object {
                          "type": "number",
                        },
                      }
        }
        value={
                Object {
                        "image": "ubuntu",
                        "scale": 3,
                      }
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "schemas": Object {
            "image": Object {
              "type": "string",
            },
            "scale": Object {
              "type": "number",
            },
          },
          "value": Object {
            "image": "ubuntu",
            "scale": 3,
          },
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <styled.div>
          <styled.div
                    onClick={[Function]}
          >
                    <styled.span>
                              config
                              
                    </styled.span>
                    <ReactImageFallback
                              fallbackImage="arrow-right.svg"
                              initialImage={null}
                              src="atom://molecule-dev-environment/.storybook/public/arrow-right.svg"
                              style={
                                        Object {
                                                  "display": "flex",
                                                  "height": "16px",
                                                  "transform": "rotate(90deg)",
                                                  "transition": "0.2s transform",
                                                  "width": "16px",
                                                }
                              }
                    />
          </styled.div>
          <styled.div>
                    <PlanConfigPart
                              name="image"
                              type="string"
                    />
                    <PlanConfigPart
                              name="scale"
                              type="number"
                    />
          </styled.div>
</styled.div>,
        "_debugID": 2,
        "_renderedOutput": <styled.div>
          <styled.div
                    onClick={[Function]}
          >
                    <styled.span>
                              config
                              
                    </styled.span>
                    <ReactImageFallback
                              fallbackImage="arrow-right.svg"
                              initialImage={null}
                              src="atom://molecule-dev-environment/.storybook/public/arrow-right.svg"
                              style={
                                        Object {
                                                  "display": "flex",
                                                  "height": "16px",
                                                  "transform": "rotate(90deg)",
                                                  "transition": "0.2s transform",
                                                  "width": "16px",
                                                }
                              }
                    />
          </styled.div>
          <styled.div>
                    <PlanConfigPart
                              name="image"
                              type="string"
                    />
                    <PlanConfigPart
                              name="scale"
                              type="number"
                    />
          </styled.div>
</styled.div>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <PlanConfigGroup
    schemas={
        Object {
            "image": Object {
              "type": "string",
            },
            "scale": Object {
              "type": "number",
            },
          }
    }
    value={
        Object {
            "image": "ubuntu",
            "scale": 3,
          }
    }
/>,
}
`;
